src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,271 @@
+    // a predicate whose support is not contained in the key columns, it
+    // must be replaced with (key, None) in case if there was previously a
+    // row with the same key that matched the predicate.
+    let key_col_len = src_type.keys[0].len();
+    let mut predicates = operator.predicates.clone();
+    let mut key_predicates = Vec::new();
+    predicates.retain(|p| {
+        let key_predicate = p.support().into_iter().all(|c| c < key_col_len);
+        if key_predicate {
+            key_predicates.push(p.clone());
+        }
+        !key_predicate
+    });
+
+    let mut storage = Vec::new();
+    let mut filtered_storage = Vec::new();
+    let mut row_packer = repr::RowPacker::new();
+
+    stream.unary_fallible(Pipeline, "UpsertLinearFallible", move |_, _| {
+        move |input, ok_output, err_output| {
+            input.for_each(|time, data| {
+                let mut ok_session = ok_output.session(&time);
+                let mut err_session = err_output.session(&time);
+                data.swap(&mut storage);
+                for (key, value, time) in storage.drain(..) {
+                    let temp_storage = RowArena::new();
+                    let datums = key.unpack();
+                    let key_pred_eval = key_predicates
+                        .iter()
+                        .map(|predicate| predicate.eval(&datums, &temp_storage))
+                        .find(|result| result != &Ok(Datum::True));
+                    match key_pred_eval {
+                        None => {
+                            if let Some(value) = value {
+                                let datums = value.unpack();
+                                let pred_eval = predicates
+                                    .iter()
+                                    .map(|predicate| predicate.eval(&datums, &temp_storage))
+                                    .find(|result| result != &Ok(Datum::True));
+                                match pred_eval {
+                                    None => {
+                                        let value = Some(row_packer.pack(position_or.iter().map(
+                                            |pos_or| match pos_or {
+                                                Some(index) => datums[*index],
+                                                None => Datum::Dummy,
+                                            },
+                                        )));
+                                        filtered_storage.push((key, value, time));
+                                    }
+                                    Some(Ok(Datum::False)) => {
+                                        filtered_storage.push((key, None, time))
+                                    }
+                                    Some(Ok(Datum::Null)) => {
+                                        filtered_storage.push((key, None, time))
+                                    }
```

Comment:
Below you use the `Some(Ok(Datum::False)) | Some(Ok(Datum::Null)) =>` framing which looks nice (and potentially avoids issues if we need to add bits to the blocks of logic). Would it make sense here too?